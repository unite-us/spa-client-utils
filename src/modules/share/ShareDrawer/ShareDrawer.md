```jsx
const groups = [
  { id: 'group-1', name: 'Group 1' },
  { id: 'group-2', name: 'Group 2' },
  { id: 'group-3', name: 'Group 3' },
  { id: 'group-4', name: 'Group 4' },
];

initialState = {
  open: false,
};

<div>
  <Button
    label={state.open ? 'Close Share Drawer' : 'Open Share Drawer'}
    onClick={() => setState({ open: !state.open })}
    primary={!state.open}
  />
  <ShareDrawer
    className="this-share-drawer"
    shareFormProps={{
      onSend: values => alert(`Send ${values.messageType} to ${values.to}`),
      sharesUrl: 'thesharesurl.com',
      id: '12a',
    }}
    groups={groups}
    onClose={() => setState({ open: !state.open })}
    open={state.open}
  />
</div>
```
