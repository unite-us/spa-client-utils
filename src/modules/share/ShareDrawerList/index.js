import ShareDrawerList from './ShareDrawerList';

export default ShareDrawerList;
