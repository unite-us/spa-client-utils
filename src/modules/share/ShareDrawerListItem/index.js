import ShareDrawerListItem from './ShareDrawerListItem';

export default ShareDrawerListItem;
