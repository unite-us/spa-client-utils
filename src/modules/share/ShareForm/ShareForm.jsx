import React, { Component } from 'react';
import { connect } from 'react-redux';
import {
  formValueSelector,
  reduxForm,
  Field,
} from 'redux-form';
import PropTypes from 'prop-types';
import {
  Button,
  InputField,
  RadioField,
} from '@unite-us/ui';
import classNames from 'classnames';
import { isError, map, noop, isFunction } from 'lodash';
import {
  emailValidation,
  phoneNumberValidation,
  requiredValidation,
} from './validations';

import InfoPopover from '../../InfoPopover';
import { SHARE_FORM_POPOVER_CONTENT } from './constants';
import { SHARE_FORM } from '../ShareForm';

const MESSAGE_TYPE_VALUES = {
  PHONE: 'sms',
  EMAIL: 'email',
  PRINT: 'print',
};

export class ShareForm extends Component {
  constructor() {
    super();
    this.onCancel = this.onCancel.bind(this);
    this.onPrint = this.onPrint.bind(this);
    this.onSend = this.onSend.bind(this);
    this.onChangeOrReset = this.onChangeOrReset.bind(this);
  }

  UNSAFE_componentWillReceiveProps(nextProps) {
    const { open, reset } = this.props;
    if (open && nextProps.open === false) {
      reset();
    }
  }

  onCancel() {
    this.props.reset();
    this.props.onCancel();
  }

  onSend(values) {
    const { email, messageType, phone } = values;
    const to = { email, sms: phone };

    return Promise.resolve(
      this.props.onSend({
        messageType,
        to: to[messageType],
      }),
    ).then((response) => {
      if (!isError(response)) {
        this.props.onTrackShare({ messageType, to: to[messageType] });
        this.onCancel();
      }
    });
  }

  onPrint(event) {
    // Prevent submission of the form.
    event.preventDefault();

    const { handlePrint, sharedGroups, sharesUrl } = this.props;
    const sharedGroupIds = map(sharedGroups, 'id').join(',');
    const url = `${sharesUrl}/?groupId=${sharedGroupIds}&print=true`;
    const win = handlePrint(url, '_blank');

    this.props.onTrackShare({ messageType: MESSAGE_TYPE_VALUES.PRINT });

    if (isFunction(this.props.onPrint)) {
      this.props.onPrint();
    }

    if (win && isFunction(win.focus)) {
      win.focus();
    }

    return win;
  }

  onChangeOrReset() {
    if (this.props.onChangeMessageType === noop) {
      this.props.reset();
    }

    this.props.onChangeMessageType();
  }

  messageTypeOptions(id) {
    return [
      {
        label: (
          <span className="sms-cont">
            Text
            <InfoPopover id={`${id}-sms-info`} placement="right" >
              <div className="share-form__popover-content">
                {SHARE_FORM_POPOVER_CONTENT}
              </div>
            </InfoPopover>
          </span>
        ),
        value: MESSAGE_TYPE_VALUES.PHONE,
      },
      { label: 'Email', value: MESSAGE_TYPE_VALUES.EMAIL },
      { label: 'Print', value: MESSAGE_TYPE_VALUES.PRINT },
    ];
  }

  render() {
    const {
      className,
      handleSubmit,
      hideCancelButton,
      id,
      messageType,
    } = this.props;
    const { EMAIL, PHONE, PRINT } = MESSAGE_TYPE_VALUES;

    return (
      <form
        className={classNames('share-form', className)}
        id={id}
        onSubmit={handleSubmit(this.onSend)}
      >
        <Field
          className="share-form__message-type-field"
          component={RadioField}
          id="share-message-type"
          inline
          label="Share Method"
          name="messageType"
          onChange={this.onChangeOrReset}
          options={this.messageTypeOptions(id)}
        />

        {messageType === PHONE &&
          <Field
            component={InputField}
            id="share-phone-field"
            label="Phone Number"
            name="phone"
            required
            validate={[requiredValidation, phoneNumberValidation]}
          />
        }

        {messageType === EMAIL &&
          <Field
            component={InputField}
            id="share-email-field"
            label="Email"
            name="email"
            required
            validate={[requiredValidation, emailValidation]}
          />
        }

        <div className="share-form__buttons">
          {!hideCancelButton &&
            <Button
              className="mr-half"
              id="share-cancel-button"
              label="Cancel"
              onClick={this.onCancel}
              secondary
            />
          }

          {messageType === PRINT ?
            <Button
              id="share-print-button"
              label="Print"
              onClick={this.onPrint}
              primary
            /> :
            <Button
              id="share-send-button"
              label="Send"
              primary
              type="submit"
            />
          }
        </div>
      </form>
    );
  }
}


export const shareFormProps = {
  /** class name */
  className: PropTypes.string,
  /** indicates location to store form in redux state, defaults to 'form' */
  formStore: PropTypes.oneOf(['form','form7']), // eslint-disable-line
  /** redux form method to connect to redux state, defaults to state => state.form */
  getFormState: PropTypes.func, // eslint-disable-line
  /** handlePrint the function used to spawn a new window with the share information, defaults to `window.open` */
  handlePrint: PropTypes.func,
  /** redux form handleSubmit function */
  handleSubmit: PropTypes.func.isRequired,
  /** hides cancel button, defaults to false */
  hideCancelButton: PropTypes.bool,
  /** id required for generating and reading unique popover id */
  id: PropTypes.string.isRequired,
  /** message type field, defaults to 'sms' */
  messageType: PropTypes.oneOf(['sms', 'email', 'print']),
  /** called on cancel */
  onCancel: PropTypes.func,
  /** called on print */
  onPrint: PropTypes.func,
  /** onChange function for message type radio field */
  onChangeMessageType: PropTypes.func,
  /** called on sending email or sms */
  onSend: PropTypes.func,
  /** tracking function called when printing/sending message */
  onTrackShare: PropTypes.func,
  /** open/closed state of drawer */
  open: PropTypes.bool,
  /** redux form reset function */
  reset: PropTypes.func.isRequired,
  /** shared groups with ids to construct complete shares url */
  sharedGroups: PropTypes.arrayOf(
    PropTypes.shape({
      id: PropTypes.string.isRequired,
    }),
  ).isRequired,
  /** base shares url to go to on print */
  sharesUrl: PropTypes.string.isRequired,
};

ShareForm.propTypes = shareFormProps;

ShareForm.defaultProps = {
  className: '',
  email: undefined,
  formStore: 'form',
  // getFormState: state => state.form, This is the default, but adding the defaultProp here does nothing
  handlePrint: window.open,
  hideCancelButton: false,
  messageType: 'sms',
  onCancel: noop,
  onPrint: noop,
  onChangeMessageType: noop,
  onSend: noop,
  onTrackShare: noop,
  open: false,
  phone: undefined,
};

function mapStateToProps(state, ownProps) {
  // Default formState value to 'form' because default prop isn't set yet at
  // this point.
  const { formStore = 'form' } = ownProps;
  const selector = formValueSelector(
    SHARE_FORM,
    reduxState => reduxState[formStore],
  );
  const messageType = selector(state, 'messageType');

  return {
    initialValues: {
      messageType: 'sms',
    },
    messageType,
  };
}

export default connect(
  mapStateToProps,
)(reduxForm({
  form: 'share',
  destroyOnUnmount: false,
})(ShareForm));
