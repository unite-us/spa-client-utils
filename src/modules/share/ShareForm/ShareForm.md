
```jsx
<div style={{ width: '300px' }}>
  <ShareForm
    id='shareform'
    initialValues={{
      messageType: 'sms',
    }}
    onSend={values => {
        setState({ result: values });
      }}
    sharedGroups={[{ id: 'shared-group-id' }]}
    sharesUrl="the.shares.url"
  />
</div>

<SandboxFormResults formResult={state.result} />

```
