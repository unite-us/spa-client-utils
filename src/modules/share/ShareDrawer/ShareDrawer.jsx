import React from 'react';
import { PropTypes } from 'prop-types';
import { Drawer } from '@unite-us/ui';
import { pluralize } from 'humanize-plus';
import classNames from 'classnames';
import { noop } from 'lodash';
import ShareDrawerList from '../ShareDrawerList';
import ShareForm from '../ShareForm';

const ShareDrawer = ({ className, groups, shareFormProps, open, ...props }) => (
  <Drawer
    className={classNames('share-drawer', className)}
    secondary
    open={open}
    {...props}
  >
    <h4 className="share-drawer__header">
      {`Share ${groups.length} ${pluralize(groups.length, 'Organization')}`}
    </h4>
    <ShareForm
      className={classNames('share-drawer__share-form', shareFormProps.className)}
      hideCancelButton
      initialValues={{
        messageType: 'sms',
      }}
      open={open}
      sharedGroups={groups}
      {...shareFormProps}
    />
    <ShareDrawerList listItems={groups} />
  </Drawer>
);

ShareDrawer.propTypes = {
  /** class name */
  className: PropTypes.string,
  /** groups to display in list */
  groups: PropTypes.arrayOf(
    PropTypes.shape({
      id: PropTypes.string.isRequired,
    }),
  ),
  /** controls drawer open/close, defaults to false */
  open: PropTypes.bool,
  /** override default ShareForm prop values. */
  shareFormProps: PropTypes.shape({
    ...ShareForm.propTypes,
    className: PropTypes.string,
    /** redux form method to connect to redux state, defaults to state => state.form */
    getFormStore: PropTypes.func,
    /** ShareForm id */
    id: PropTypes.string.isRequired,
    /** called on send email/phone */
    onSend: PropTypes.func,
    /** tracking call made sharing group data */
    onTrackShare: PropTypes.func,
    /** base url for shares app */
    sharesUrl: PropTypes.string.isRequired,
  }).isRequired,
  /** drawer width, defaults to 380px */
  width: PropTypes.string,
};

ShareDrawer.defaultProps = {
  className: '',
  groups: [],
  open: false,
  shareFormProps: {
    onSend: noop,
    onTrackShare: noop,
  },
  width: '380px',
};

export default ShareDrawer;
