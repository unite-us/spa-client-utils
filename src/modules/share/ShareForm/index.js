import ShareForm from './ShareForm';

export const SHARE_FORM = 'share';
export default ShareForm;
