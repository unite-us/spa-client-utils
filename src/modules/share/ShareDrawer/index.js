import ShareDrawer from './ShareDrawer';

export default ShareDrawer;
