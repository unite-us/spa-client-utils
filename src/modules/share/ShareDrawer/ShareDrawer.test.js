import React from 'react';
import renderer from 'react-test-renderer';
import ShareDrawer from './ShareDrawer';

require('modules/share/ShareForm').default = jest.fn(() => <mock-share-form />);

const groups = [
  { id: 'group-1', name: 'Group 1' },
  { id: 'group-2', name: 'Group 2' },
  { id: 'group-3', name: 'Group 3' },
  { id: 'group-4', name: 'Group 4' },
];

const props = {
  groups,
  shareFormProps: {
    onSend: jest.fn(),
    sharesUrl: 'the.shares.url.com',
    id: 'test-id',
  },
};

const singleGroupProps = {
  groups: [{ id: 'group-1', name: 'Group 1' }],
  shareFormProps: {
    onSend: jest.fn(),
    sharesUrl: 'the.shares.url.com',
    id: 'test-id',
  },
};

describe('ShareDrawer', () => {
  it('renders a closed drawer', () => {
    expect(renderer.create(
      <ShareDrawer {...props} />,
    )).toMatchSnapshot();
  });

  it('renders an open drawer with header and group list', () => {
    expect(renderer.create(
      <ShareDrawer {...props} open />,
    )).toMatchSnapshot();
  });

  it('renders an open drawer with correct header based on group length', () => {
    expect(renderer.create(
      <ShareDrawer {...singleGroupProps} open />,
    )).toMatchSnapshot();
  });
});
