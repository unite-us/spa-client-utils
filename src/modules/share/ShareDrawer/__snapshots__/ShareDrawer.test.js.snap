// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ShareDrawer renders a closed drawer 1`] = `
<div
  className="ui-drawer ui-drawer--secondary share-drawer"
  id=""
  style={
    Object {
      "height": "calc(100% - 0)",
      "top": "0",
      "width": "380px",
    }
  }
>
  <div
    className="ui-drawer__container"
  >
    <div
      className="ui-drawer__body"
    >
      <div
        className="ui-drawer__title"
      >
        <div
          className="ui-drawer__title__close-button"
        >
          <a
            aria-label="Back"
            onClick={[Function]}
            onKeyDown={[Function]}
            role="button"
            tabIndex={0}
          >
            <svg
              className="ui-icon"
              height="16px"
              id=""
              style={
                Object {
                  "display": "inline-block",
                  "fill": "#506784",
                  "verticalAlign": "middle",
                }
              }
              viewBox="0 0 24 32"
              width="16px"
            >
              <path
                d="M20.911 5.375l-9.482 9.482 9.482 9.482q0.339 0.339 0.339 0.804t-0.339 0.804l-2.964 2.964q-0.339 0.339-0.804 0.339t-0.804-0.339l-13.25-13.25q-0.339-0.339-0.339-0.804t0.339-0.804l13.25-13.25q0.339-0.339 0.804-0.339t0.804 0.339l2.964 2.964q0.339 0.339 0.339 0.804t-0.339 0.804z"
              />
            </svg>
          </a>
        </div>
      </div>
      <h4
        className="share-drawer__header"
      >
        Share 4 Organizations
      </h4>
      <mock-share-form />
      <ul
        className="share-drawer-list"
      >
        <li
          className="share-drawer-list-item"
        >
          Group 1
        </li>
        <li
          className="share-drawer-list-item"
        >
          Group 2
        </li>
        <li
          className="share-drawer-list-item"
        >
          Group 3
        </li>
        <li
          className="share-drawer-list-item"
        >
          Group 4
        </li>
      </ul>
    </div>
    <div
      className="ui-drawer__close-btn share-drawer"
    >
      <a
        aria-label="Close"
        onClick={[Function]}
        onKeyDown={[Function]}
        role="button"
        tabIndex={0}
      >
        <svg
          className="ui-icon"
          height="14px"
          id=""
          style={
            Object {
              "display": "inline-block",
              "fill": "#B7D2E5",
              "verticalAlign": "middle",
            }
          }
          viewBox="0 0 21 21"
          width="14px"
        >
          <g>
            <path
              d="M2.592.044l18.364 18.364-2.548 2.548L.044 2.592z"
            />
            <path
              d="M0 18.364L18.364 0l2.548 2.548L2.548 20.912z"
            />
          </g>
        </svg>
      </a>
    </div>
  </div>
</div>
`;

exports[`ShareDrawer renders an open drawer with correct header based on group length 1`] = `
<div
  className="ui-drawer ui-drawer--secondary ui-drawer--opened share-drawer"
  id=""
  style={
    Object {
      "height": "calc(100% - 0)",
      "top": "0",
      "width": "380px",
    }
  }
>
  <div
    className="ui-drawer__container"
  >
    <div
      className="ui-drawer__body"
    >
      <div
        className="ui-drawer__title"
      >
        <div
          className="ui-drawer__title__close-button"
        >
          <a
            aria-label="Back"
            onClick={[Function]}
            onKeyDown={[Function]}
            role="button"
            tabIndex={0}
          >
            <svg
              className="ui-icon"
              height="16px"
              id=""
              style={
                Object {
                  "display": "inline-block",
                  "fill": "#506784",
                  "verticalAlign": "middle",
                }
              }
              viewBox="0 0 24 32"
              width="16px"
            >
              <path
                d="M20.911 5.375l-9.482 9.482 9.482 9.482q0.339 0.339 0.339 0.804t-0.339 0.804l-2.964 2.964q-0.339 0.339-0.804 0.339t-0.804-0.339l-13.25-13.25q-0.339-0.339-0.339-0.804t0.339-0.804l13.25-13.25q0.339-0.339 0.804-0.339t0.804 0.339l2.964 2.964q0.339 0.339 0.339 0.804t-0.339 0.804z"
              />
            </svg>
          </a>
        </div>
      </div>
      <h4
        className="share-drawer__header"
      >
        Share 1 Organization
      </h4>
      <mock-share-form />
      <ul
        className="share-drawer-list"
      >
        <li
          className="share-drawer-list-item"
        >
          Group 1
        </li>
      </ul>
    </div>
    <div
      className="ui-drawer__close-btn ui-drawer__close-btn--opened share-drawer"
    >
      <a
        aria-label="Close"
        onClick={[Function]}
        onKeyDown={[Function]}
        role="button"
        tabIndex={0}
      >
        <svg
          className="ui-icon"
          height="14px"
          id=""
          style={
            Object {
              "display": "inline-block",
              "fill": "#B7D2E5",
              "verticalAlign": "middle",
            }
          }
          viewBox="0 0 21 21"
          width="14px"
        >
          <g>
            <path
              d="M2.592.044l18.364 18.364-2.548 2.548L.044 2.592z"
            />
            <path
              d="M0 18.364L18.364 0l2.548 2.548L2.548 20.912z"
            />
          </g>
        </svg>
      </a>
    </div>
  </div>
</div>
`;

exports[`ShareDrawer renders an open drawer with header and group list 1`] = `
<div
  className="ui-drawer ui-drawer--secondary ui-drawer--opened share-drawer"
  id=""
  style={
    Object {
      "height": "calc(100% - 0)",
      "top": "0",
      "width": "380px",
    }
  }
>
  <div
    className="ui-drawer__container"
  >
    <div
      className="ui-drawer__body"
    >
      <div
        className="ui-drawer__title"
      >
        <div
          className="ui-drawer__title__close-button"
        >
          <a
            aria-label="Back"
            onClick={[Function]}
            onKeyDown={[Function]}
            role="button"
            tabIndex={0}
          >
            <svg
              className="ui-icon"
              height="16px"
              id=""
              style={
                Object {
                  "display": "inline-block",
                  "fill": "#506784",
                  "verticalAlign": "middle",
                }
              }
              viewBox="0 0 24 32"
              width="16px"
            >
              <path
                d="M20.911 5.375l-9.482 9.482 9.482 9.482q0.339 0.339 0.339 0.804t-0.339 0.804l-2.964 2.964q-0.339 0.339-0.804 0.339t-0.804-0.339l-13.25-13.25q-0.339-0.339-0.339-0.804t0.339-0.804l13.25-13.25q0.339-0.339 0.804-0.339t0.804 0.339l2.964 2.964q0.339 0.339 0.339 0.804t-0.339 0.804z"
              />
            </svg>
          </a>
        </div>
      </div>
      <h4
        className="share-drawer__header"
      >
        Share 4 Organizations
      </h4>
      <mock-share-form />
      <ul
        className="share-drawer-list"
      >
        <li
          className="share-drawer-list-item"
        >
          Group 1
        </li>
        <li
          className="share-drawer-list-item"
        >
          Group 2
        </li>
        <li
          className="share-drawer-list-item"
        >
          Group 3
        </li>
        <li
          className="share-drawer-list-item"
        >
          Group 4
        </li>
      </ul>
    </div>
    <div
      className="ui-drawer__close-btn ui-drawer__close-btn--opened share-drawer"
    >
      <a
        aria-label="Close"
        onClick={[Function]}
        onKeyDown={[Function]}
        role="button"
        tabIndex={0}
      >
        <svg
          className="ui-icon"
          height="14px"
          id=""
          style={
            Object {
              "display": "inline-block",
              "fill": "#B7D2E5",
              "verticalAlign": "middle",
            }
          }
          viewBox="0 0 21 21"
          width="14px"
        >
          <g>
            <path
              d="M2.592.044l18.364 18.364-2.548 2.548L.044 2.592z"
            />
            <path
              d="M0 18.364L18.364 0l2.548 2.548L2.548 20.912z"
            />
          </g>
        </svg>
      </a>
    </div>
  </div>
</div>
`;
